### ISSUE TITLE
caching with dependencies on files

### ISSUE THREAD

**Issue by barryrowlingson** [yihui/knitr#238]
If I have a chunk that reads a data file then modifying the data file doesn't invalidate the cached chunk. Obviously it would be hard for build_dep to spot these, but maybe they could be specified in the chunk header? Something like:

```
<<data, fdep="foo.csv">>
data = read.csv("foo.csv")
@ 
```

then this chunk would be fetched from the cache unless the file foo.csv has changed.  Not sure how you'd specify multiple dependent data files on the chunk header...

Anyway, caching has already saved me enough time to submit this as a feature request :)

**Comment by zachary-foster** [comment]
This is great for checking for input file changes! However, sometimes I want to rerun a cached chunk when the **output** of that chunk has changed. For example, if an output file needed for following chunks is deleted, it would be nice if that chunk would detect the lack of the output file and recreate it next time the Rmd is built.

I might be missing somthing obvious, but I have tried for a while, and have not found a way to use `cache.extra` to accomplish this. Using  `cash.extra=mtime(output_files)` does not work since the the chunk changes the modification times of the output files. Using  `cash.extra=file.exists(output_files)` almost works, but it has to be run twice; The first run after a file is deleted  `cash.extra=FALSE` and the chunk is run, creating output files, so next time `cash.extra=TRUE` and the chunk is run again. After that the chunk is not run anymore.  

The best working solution I have so far is to delete the cache when output files are not found. 

```{r}
delete_chunk_cache <- function(chunk_name) {
  cache_path <- opts_current$get("cache.path")
  cached_files <- list.files(cache_path)
  search_pattern <- paste0("^", chunk_name, "_[a-z0-9]+\\.[a-z0-9]+$")
  chunk_cache_files <- cached_files[grepl(search_pattern, cached_files)]
  file.remove(file.path(cache_path, chunk_cache_files))
}

delete_next_chunk_cache <- function() {
  next_chunk <- all_labels()[which(all_labels() == opts_current$get("label")) + 1]
  delete_chunk_cache(next_chunk)
}

mtime <- function(files){
  lapply(Sys.glob(files),function(x) file.info(x)$mtime)
}
```

```{r}

```{r chunk_1}
input_files = c("list", "of", "paths")
output_files = c("list", "of", "paths")
if (any(!file.exists(output_files))) delete_next_chunk_cache()
```

```{r}

```{r chunk_2, cache=TRUE, cash.extra=mtime(input_files)}
system(some_command_that_makes_output_files)
```
Do you know of a better way to do this? Thanks!

**Comment by yihui** [comment]
@zachary-foster You can probably disable cache if the file does not exist, e.g. `cache = file.exists('some_file')`.

**Comment by zachary-foster** [comment]
Thanks for the quick reply! I tried that and it works ok, but it has one small issue. If the output file is deleted and a change is made to the code, it takes two runs of the chunk to cache correctly. On the first build after the deletion of the output file and a change in code, `cache=FALSE` so it runs (like it should), but it does not record the change in code, so when it is built a second time and `cache=TRUE` (since output files were created on the first run) it notices the change in code and runs again.

However, a simultaneous change in output files and code is probably rare enough that this solution is sufficient. 

Would a chunk option like `rebuild.cache = !file.exists('some_file')` be useful? `rebuild.cache` would not disable caching, but when `TRUE`, it forces the chunk to be evaluated regardless of code changes. I think that would work right if the code and output are changed at the same time. I do a bit of R package development. I can try to add the `rebuild.cache` chunk option if you think it would be useful. 

Thank you!

**Comment by yihui** [comment]
@zachary-foster Normally caching invalidation is only about the _input_: if the input changes, we invalidate the cache. In your case, it sounds there is a circular dependency -- the code chunk depends on its _output_. I'm not sure why that has to be the case, i.e. why this code chunk needs its own output file to run.

**Comment by zachary-foster** [comment]
It does not need its own output file to run, but it is part of a "pipeline" that calls many external commands. During a given build, it is checking for the ouput of the previous build. If the chunk's output file is deleted, but the code is not changed, the chunk is not run and then the following chunks break because the commands they call need the first chunks output.

**Comment by yihui** [comment]
@zachary-foster I see. That makes sense now. A pull request is welcome :) I prefer the option name to be `cache.rebuild` instead of `rebuild.cache`. When you calculate the MD5 digest, remember to exclude this option: https://github.com/yihui/knitr/blob/master/R/block.R#L81

**Comment by zachary-foster** [comment]
Thanks for the tip! That probably saved me some debugging time. Pull request sent. Thanks for considering it.

**Comment by jayrobwilliams** [comment]
Just a quick update to @jimhester's suggestion above. I don't know if something low level has changed in R in the past six years, but his `mtime` function failed when I tried to feed it multiple filenames.

I had to make just a quick tweak to get it to accept multiple filenames as separate arguments:

```r
mtime <- function(file, ...) {
  lapply(Sys.glob(c(file, ...)), function(x) file.info(x)$mtime)
}
```

Hopefully this is helpful for anyone that stumbles on this issue in the future.
